mkosi/cage.py

```diff
@@ -0,0 +1,667 @@
+    root within it. The kernel only allows a process running outside of the unshared user namespace to write the
+    necessary uid and gid mappings, so we fork off a child process, make it wait until the parent process has unshared
+    a user namespace, and then writes the necessary uid and gid mappings.
+    """
+    ppid = os.getpid()
+
+    event = libc.eventfd(0, 0)
+    if event < 0:
+        oserror()
+
+    pid = os.fork()
+    if pid == 0:
+        try:
+            os.read(event, ctypes.sizeof(ctypes.c_uint64))
+            os.close(event)
+            with open(f"/proc/{ppid}/setgroups", "w") as f:
+                f.write("deny\n")
+            with open(f"/proc/{ppid}/gid_map", "w") as f:
+                f.write(f"{gid} {os.getgid()} 1\n")
+            with open(f"/proc/{ppid}/uid_map", "w") as f:
+                f.write(f"{uid} {os.getuid()} 1\n")
+        except OSError as e:
+            os._exit(e.errno)
+        except BaseException:
+            os._exit(1)
+        else:
+            os._exit(0)
+
+    try:
+        unshare(CLONE_NEWUSER)
+    finally:
+        os.write(event, ctypes.c_uint64(1))
+        os.close(event)
+        _, status = os.waitpid(pid, 0)
+
+    rc = os.waitstatus_to_exitcode(status)
+    if rc != 0:
+        raise OSError(rc, os.strerror(rc))
+
+
+def acquire_privileges(*, become_root: bool = False) -> bool:
+    if os.getuid() == 0 or (not become_root and have_effective_cap(CAP_SYS_ADMIN)):
+        return False
+
+    become_user(0 if become_root else os.getuid(), 0 if become_root else os.getgid())
+
+    if not become_root:
+        cap_permitted_to_ambient()
+
+    return True
+
+
+def userns_has_single_user() -> bool:
+    if not os.path.exists("/proc/self/uid_map"):
+        return False
+
+    with open("/proc/self/uid_map") as f:
+        lines = f.readlines()
+
+    return len(lines) == 1 and int(lines[0].split()[-1]) == 1
+
+
+def resolve(root: str, path: str) -> str:
+    fd = os.open("/", os.O_CLOEXEC|os.O_PATH|os.O_DIRECTORY)
+
+    try:
+        os.chroot(root)
+        os.chdir("/")
+        return os.path.join(root, os.path.realpath(path).lstrip("/"))
+    finally:
+        os.fchdir(fd)
+        os.close(fd)
+        os.chroot(".")
+
+
+def splitpath(path: str) -> tuple[str, ...]:
+    return tuple(p for p in path.split("/") if p)
+
+
+def pop_all(list: list[str]) -> list[str]:
+    result = []
+
+    while list:
+        result.append(list.pop())
+
+    return result
+
+
+def do_mount(operation: str, args: tuple[str, ...]) -> None:
+    src, dst = args
+    src = resolve("oldroot", src)
+    dst = resolve("newroot", dst)
+
+    if not os.path.exists(src) and operation.endswith("try"):
+        return
+
+    with umask(~0o755):
+        os.makedirs(os.path.dirname(dst), exist_ok=True)
+
+    if not os.path.exists(dst):
+        with umask(~0o644 if os.path.isfile(src) else ~0o755):
+            if os.path.isfile(src):
+                os.close(os.open(dst, os.O_CREAT|os.O_CLOEXEC))
```

Comment:
Let's hoist the check up a level
```suggestion
        isfile = os.path.isfile(src)
        with umask(~0o644 if isfile else ~0o755):
            if isfile:
                os.close(os.open(dst, os.O_CREAT|os.O_CLOEXEC))
```